### PR TITLE
api: allow manifests to be pulled by digest (PROJQUAY-5467)

### PR DESCRIPTION
**Pull request by Marcusk19**
Prevent manifests pushed by digest from being hidden. Lets newer versions of conftest push to quay, since it expects to be able to pull the policy by digest when performing a push.

### CODE REVIEW

**Comment by flavianmissi** [comment]
The need behind this change is unclear in this PR's description - would you mind expanding a little more on that @Marcusk19 ?

File: endpoints/v2/manifest.py
Comment (dmage) [inline endpoints/v2/manifest.py:149]: allow_dead isn't exactly what you need. It'll return expired manifests, which is incorrect. Quay should always pretend that they are deleted.

Why do you need `require_available`? What happens if `require_available` isn't set?

**Review by dmage — CHANGES_REQUESTED** [review]
We need tests that confirm that this feature works as expected and Quay doesn't accidentally distribute expired images.

**Comment by Marcusk19** [comment]
@flavianmissi Yeah, this is supposed to address an error you get when [pushing an OPA policy to quay with conftest](https://www.conftest.dev/sharing/#pushing-to-an-oci-registry):
```
Error: push bundle: tagging: 
sha256:fae4613127bc35df711682c11e462dd9db2d2716dca45553c00669457eda91b1: not found 
```
It looks like this error occurs because manifests pushed by their digest are hidden by default, and conftest expects to be able to get that manifest to verify that it exists. This check was introduced in newer versions of conftest, which is why we didn't run into this issue in the past.

**Comment by Marcusk19** [comment]
@dmage Would the already existing registry tests cover this? It looks like there's [test_basic_push_by_manifest_digest](https://github.com/quay/quay/blob/d49dbd1515b3a17d21127b881461d333045552a5/test/registry/registry_tests.py#L204) that can be used to test pulling by digest and [test_attempt_pull_by_manifest_digest_for_deleted_tag](https://github.com/quay/quay/blob/d49dbd1515b3a17d21127b881461d333045552a5/test/registry/registry_tests.py#L2834) for testing expired images aren't returned

**Comment by flavianmissi** [comment]
good job! lgtm but let's wait for others' feedback before landing this

**Comment by dmage** [comment]
It makes temporary tags for manifests pushed by digest visible in UI.

Screenshot 2023-05-24 at 12 41 20

**Comment by flavianmissi** [comment]
> @dmage I saw the discussion on slack mentioning using the digest to tag manifests pushed by digest. Should we do that instead of using these temp tags?

I like this solution.

@Marcusk19 I see you added a new flag for tagging by digest, is there a reason to not ditch the $temp-.* stuff completely? Will we break anything by doing that? 🤔

**Comment by flavianmissi** [comment]
Unrelatedly, it looks like some of your tests are failing, please investigate if the failure is related to your change as it might affect the solution 🙏🏼

**Comment by Marcusk19** [comment]
@flavianmissi I think we're going back to the original approach, but instead of `allow_dead=True` a new boolean called `allow_hidden` is used to return the hidden but not expired tags. Should prevent the UI from becoming noisy with all the temp tags.

**Review by dmage — APPROVED** [review]

**Review by flavianmissi — APPROVED** [review]

**Pull request merged**
